daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/parsers/DelimiterParsers.scala

```diff
@@ -51,27 +51,22 @@ class DelimiterTextParser(
 
   override val nom = delimiterType.toString
 
-  private def containsLocalMatch(delimiters: ArrayBuffer[DFADelimiter], state: PState): Boolean = {
-    var i = delimiters.length - 1
+  private def containsLocalMatch(matchedDelimiters: ArrayBuffer[DFADelimiter], state: PState): Boolean = {
+    val localIndexStart = state.mpstate.delimitersLocalIndexStack.top
+    val inScopeDelimiters = state.mpstate.delimiters
+
+    var foundLocalDFAIndex = -1
+
     // scan matches in reverse. local matches are going to be at the end
-    while (i >= 0) {
-      val dfa = delimiters(i)
-      if (dfa.indexInDelimiterStack < state.mpstate.delimitersLocalIndexStack.top) {
-        // we found a remote delim. since all remote delimiters are first, and
-        // we are traversing this in reverse, we know all remaining delims are
-        // remotes, so we can return early
-        return false
-      }
-      // note, it is possible for dfaIndex to be greater than the length of the
-      // found delimiters. This just means a found delimiter went out of scope
-      // (and thus, not a local delimiter)
-      if (dfa.delimType == delimiterType && dfa.indexInDelimiterStack < state.mpstate.delimiters.length) {
-        // found a local delim
-        return true
+    var i = matchedDelimiters.length - 1
+    while (foundLocalDFAIndex < 0 && i >= 0) {
+      val dfa = matchedDelimiters(i)
+      if (dfa.delimType == delimiterType) {
```

Comment:
We should consider constructing a really pathological schema with nest after nest after nest of separated sequences so that this while loop actually has to search down a bit. Just so we can understand how bad it has to be before this matters at all.  It's possible that for an array, where the same delimiter is being searched repeatedly that this loop does not need to iterate anew for each one. But that's an optimization I would put in only if we had measurements.

Reply:
I did as you suggested and created a schema with deeply nested separators and some pretty ugly data that would find lots of remote delimiters and requiring lots of scanning of the delimter stack. This test didn't find any noticeable performance differences. At most it was just 1-2% slower, but that's within the margin of error for java, and I had test runs that were just as performant as before this change.

I also inserted some time trackers, and it looks like only about 6% of the total parse time is spent in the parts of  DelimiterStackParser that isn't doing actual delimiter scanning (so that includes creating delimiter iterators, determining local/remote which is what this change affects, skipping matched characters, creating ParseErrors, etc). 41% of the parse is spent in the actual delimiter scanning part, and 30% in the sequence parser stuff, and miscellaneous other stuff.

So I think this index stuff was just a premature optimization--the bulk of time is spent elsewhere.